airbyte-db/lib/src/main/java/io/airbyte/db/Database.java

```diff
@@ -4,48 +4,89 @@
 
 package io.airbyte.db;
 
-import java.io.Closeable;
+import io.airbyte.commons.lang.Exceptions;
+import java.io.IOException;
 import java.sql.SQLException;
-import javax.sql.DataSource;
-import org.jooq.SQLDialect;
+import java.util.function.Function;
+import lombok.val;
+import org.jooq.DSLContext;
 import org.jooq.impl.DSL;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Database object for interacting with a Jooq connection.
  */
 public class Database implements AutoCloseable {
 
-  private final DataSource ds;
-  private final SQLDialect dialect;
+  private static final Logger LOGGER = LoggerFactory.getLogger(Database.class);
 
-  public Database(final DataSource ds, final SQLDialect dialect) {
-    this.ds = ds;
-    this.dialect = dialect;
+  private static final long DEFAULT_WAIT_MS = 5 * 1000;
+
+  private final DSLContext dslContext;
+
+  public Database(final DSLContext dslContext) {
+    this.dslContext = dslContext;
   }
 
   public <T> T query(final ContextQueryFunction<T> transform) throws SQLException {
-    return transform.query(DSL.using(ds, dialect));
+    return transform.query(dslContext);
   }
 
   public <T> T transaction(final ContextQueryFunction<T> transform) throws SQLException {
-    return DSL.using(ds, dialect).transactionResult(configuration -> transform.query(DSL.using(configuration)));
-  }
-
-  public DataSource getDataSource() {
-    return ds;
+    return dslContext.transactionResult(configuration -> transform.query(DSL.using(configuration)));
   }
 
   @Override
   public void close() throws Exception {
-    // Just a safety in case we are using a datasource implementation that requires closing.
-    // BasicDataSource from apache does since it also provides a pooling mechanism to reuse connections.
+    dslContext.close();
```

Comment:
There is ambiguous ownership here that will likely cause improper resource management. Can you pass the same dslContext to multiple databases? If so then the database shouldn't be responsible for closing it. If not, maybe you can pass in a Provider.

It seems like the calling App site is also taking care of closing the context.

Reply:
@michel-tricot Yes.  I wasn't 100% sure why we have this close method in the first place.  The original implementation closed the underlying data source, which seems odd as the connection pool should manage the connections.  Once we switch over to using an application framework, we will be out of this business entirely.  

As you correctly noticed, I added shutdown handlers to the apps themselves to mimic the behavior of allowing the application framework to manage closing the underlying resource.  I'm more than happy to remove this method entirely from the `Database` class if we are comfortable with the shutdown handlers doing this for us.  From what I have seen in the the `DSLContext` code, it does properly release the connection back to the data source connection pool.  Therefore, I think we are safe to remove this method and just let the application handle closing the datasource on shutdown.